feat(TextEditor): add new component

TomasSlama opened the pull request:
<!---
Thanks for taking the time to contribute!

Please make sure to follow contribution process:
https://toptal-core.atlassian.net/wiki/spaces/FE/pages/2396094469/Handling+external+contribution
-->

[FX-2293]

### Description

Describe the changes and motivations for the pull request.

### How to test

- FIXME: Add the steps describing how to verify your changes

### Review

- [ ] Read [CONTRIBUTING.md](https://github.com/toptal/picasso/blob/master/CONTRIBUTING.md) and [Component API principles](https://github.com/toptal/picasso/blob/master/docs/contribution/component-api.md)
- [ ] Annotate all `props` in component with documentation
- [ ] Create `examples` for component
- [ ] Ensure that deployed demo has expected results and good examples
- [ ] Ensure that tests pass by running `yarn test`
- [ ] Ensure that visuals tests pass by running `yarn test:visual`. If not - check the documentation [how to fix visual tests](https://github.com/toptal/picasso/blob/master/docs/contribution/visual-testing.md#fixing-broken-visual-tests-inside-a-pr)
- [ ] Ensure the changed/created components have not caused accessibility issues. [How to use accessibility plugin in storybook](https://github.com/toptal/picasso/blob/master/docs/contribution/accessibility.md).

<details>
<summary>PR commands</summary>
<br />

List of available commands:

- `@toptal-bot run all` - Run whole pipeline
- `@toptal-bot run build` - Check build
- `@toptal-bot run visual` - Run visual tests
- `@toptal-bot run deploy:documentation` - Deploy documentation
- `@toptal-bot run package:alpha-release` - Release alpha version

</details>

<details>
<summary>PR Review Guidelines</summary>
<br />

#### When to approve? ✅

**You are OK** with merging this PR and
1. You have no extra requests.
2. You have optional requests.
   1. Add `nit: ` to your comment. (ex. `nit: I'd rename this variable from makeCircle to getCircle`)

#### When to request changes? ❌

**You are not OK** with merging this PR because
1. Something is broken after the changes.
2. Acceptance criteria is not reached.
3. Code is dirty.

#### When to comment (neither ✅ nor ❌)

**You want your comments to be addressed** before merging this PR in cases like:
1. There are leftovers like unnecessary logs, comments, etc.
2. You have an opinionated comment regarding the code that requires a discussion.
3. You have questions.

#### How to handle the comments?

1. An owner of a comment is the only one who can resolve it.
2. An owner of a comment must resolve it when it's addressed.
3. A PR owner must reply with ✅ when a comment is addressed.

</details>

TomasSlama commented on packages/picasso-lab/src/TextEditor/hooks/useHandleAutofocus.tsx:
```diff
@@ -0,0 +1,23 @@
+
+  useEffect(() => {
+    if (firstRender.current && editorRef.current && autofocus) {
+      // without setTimeout is focus called on last instance🤷‍♂️
+      setTimeout(() => {
+        editorRef.current.focus()
+        firstRender.current = false
+      }, 0)
```
For some unknown reason, when we have multiple instances of editor, even with different id, without setTimeout was the `focus()` function called on wrong instance.

TomasSlama commented on packages/picasso-lab/src/TextEditor/hooks/useHandleChangeFromController.tsx:
```diff
@@ -0,0 +1,31 @@
+
+      // is value set from outside
+      if (editorValue !== controllerValue) {
+        // TODO discuss with others, if we should sanitize here
+        // usecase - editing current job description
+        clipboard.dangerouslyPasteHTML(controllerValue)
```
since `value` needs to be always HTML string, I'm not sure if we should rely on users to sanitize the value if they decide to change from outside of the editor. Or if we should sanitize it here by ourselves. Better safe than sorry. WDYT?

teimurjan replied:
Isn't it sanitized by Quill?

TomasSlama replied:
nope 👉 [docs](https://quilljs.com/docs/modules/clipboard/#dangerouslypastehtml)

TomasSlama replied:
Teimur found better solution to solve this issue. We convert HTML back to quill delta.

teimurjan commented on packages/picasso-lab/package.json:
```diff
@@ -32,10 +32,12 @@
   },
   "dependencies": {
     "@toptal/picasso-provider": "0.4.1",
+    "@types/quill": "^2.0.9",
```
It should be in `devDependencies`, right?

teimurjan commented on packages/picasso-lab/src/TextEditor/TextEditor.tsx:
```diff
@@ -0,0 +1,95 @@
+   * number of characters.
+   */
+  // TODO implement
+  maxlength?: number
```
Not sure about this one. I've never seen a WYSIWYG editor with a length limit. It'd be better to ask BASE about it.

TomasSlama replied:
@teimurjan I asked Milos, TextEditor will be used instead of textarea for the job description, and they are restricted with max length right now and it would be great to have it on TextEditor too.

teimurjan commented on packages/picasso-lab/src/TextEditor/hooks/useHandleAutofocus.tsx:
```diff
@@ -0,0 +1,25 @@
+    const editorApi = editorRef?.current
+
+    if (firstRender?.current && editorApi && autofocus) {
+      // without setTimeout is focus called on last instance🤷‍♂️
+      setTimeout(() => {
```
Can you clarify this one a bit? Extraction of a code piece into the timer queue is not a good solution. I think we'll be able to find a more elegant one.

TomasSlama replied:
yeah, it would be great, I was struggling a bit with this one. The problem is, when you have two stories and you add `autofocus` on the first one, it will be triggered on the second one 🤷. I didn't come up with any idea why is that happening. But I will try to dig deeper.

teimurjan replied:
https://github.com/toptal/picasso/pull/2317

teimurjan replied:
🔝 it solves the unsafe HTML problem as well

teimurjan commented on packages/picasso-lab/src/TextEditor/hooks/useHandleChangeEvent.tsx:
```diff
@@ -0,0 +1,28 @@
+  { onChange }: { onChange: Props['onChange'] }
+) => {
+  useEffect(() => {
+    const { current: editorApi } = editorRef
```
We have `editorRef` meaning it contains `editor`, not `editorApi`. I'd rename `editorApi` to `editor`.

teimurjan commented on packages/picasso-lab/src/TextEditor/styles.ts:
```diff
@@ -0,0 +1,8 @@
+export default () =>
+  createStyles({
+    root: {
+      height: 'calc(100% - 43px)' // TODO proper style editor
```
Are you going to fix it in this PR?

TomasSlama replied:
@teimurjan no, we have another ticket for styling the editor. This was just a quick fix so it looks good in story.

LashaJini commented on packages/picasso-lab/src/TextEditor/TextEditor.tsx:
```diff
@@ -0,0 +1,95 @@
+
+type HTMLString = string
+
+export interface Props extends BaseProps {
```
Tbh, this confuses me. Will this text editor have only `HTML` as input, or it can also have `markdown` (in the future).

TomasSlama replied:
hm, `HTML` will be our only supported output.

Input should be any copy pasted text or written text. Programaticly you can set value with `HTML` (when you paste the output back to editor)

LashaJini commented:
When I visit texteditor component in picasso it jumps to the second example, is this intended?

![simplescreenrecorder-2021-12-09_13 44 17](https://user-images.githubusercontent.com/72510037/145372972-b5182c3a-1dde-49da-983d-0bbe2348a9da.gif)

TomasSlama commented on jest.config.js:
```diff
@@ -22,6 +22,7 @@ module.exports = {
   moduleNameMapper: {
     '\\.(jpg|jpeg|png|gif|webp|svg|ttf|woff|woff2)$':
       '<rootDir>/__tests__/fileMock.js',
+    '\\.css$': '<rootDir>/__tests__/styleMock.js',
```
unit test stopped working because of using `import quill.css`

dmaklygin1 commented on packages/picasso-lab/src/TextEditor/hooks/useDisableEditor.tsx:
```diff
@@ -0,0 +1,17 @@
+    if (editorRef.current) {
+      editorRef.current.enable(!disabled)
+    }
+  }, [disabled, editorRef])
```
if I'm not wrong, the `editorRef` won't be changed, it's the initial instance of the ref. But `editorRef.current` might be changed

TomasSlama replied:
actually the ref will never trigger the rerender for us, so it is there only to remove eslint warning. It would work the same with or without `editorRef`. 
[Interesting article on this issue](https://epicreact.dev/why-you-shouldnt-put-refs-in-a-dependency-array/)

teimurjan replied:
Whenever you change `.current` [the pointer to the ref is also updated](https://codesandbox.io/s/hardcore-chaum-8ozth?file=/src/App.js).

teimurjan replied:
My fault. When you set `.current`, it doesn't update the pointer. Thank you @dmaklygin1

TomasSlama replied:
![image](https://user-images.githubusercontent.com/6830426/145979176-471deea5-e8e0-4dbd-bf11-4a36ea6a0ad3.png)

guys, @teimurjan, @dmaklygin1, it doesn't matter if I pass `ref` or `ref.current` to `useEffect` in dependency list, changes in ref will never trigger rerender. I pass `ref` there, just to remove eslint warning. It is common issue, when we pass ref to custom hook. If it would not be in custom hook, eslint would not throw the warning. Detail explanation is in [this article](Interesting article on this issue).

And TBH I don't want to pass only mutable part of ref to hooks, I think it is much better experience to pass whole `ref` around as we do in normal components.

dmaklygin1 commented on packages/picasso-lab/src/TextEditor/hooks/useDisableEditor.tsx:
```diff
@@ -0,0 +1,17 @@
+
+const useDisableEditor = (
+  editorRef: EditorRefType,
+  { disabled }: { disabled: Props['disabled'] }
```
I'm thinking...I don't know whether it's a good experience to borrow interface from the parent's level. 
Here, we definitely know that `disabled` is boolean, shouldn't we directly describe it here? Wdyt?

OndrejTuma replied:
I'm wondering why second param is an object here 🤔 
I don't see any other purpose for this hook other than setting true/false to quill `enable` method
(similar to `useHandleAutofocus`)
Do you have some other plan to these hooks?

TomasSlama replied:
✅

TomasSlama commented:
> When I visit texteditor component in picasso it jumps to the second example, is this intended?
> 
> ![simplescreenrecorder-2021-12-09_13 44 17](https://user-images.githubusercontent.com/72510037/145372972-b5182c3a-1dde-49da-983d-0bbe2348a9da.gif)

Good catch @LashaJini, it should work now properly

dmaklygin1 commented on packages/picasso-lab/src/TextEditor/hooks/useHandleChangeEvent.tsx:
```diff
@@ -0,0 +1,28 @@
+
+    if (editor) {
+      const handler: TextChangeHandler = () => {
+        onChange(editor.root.innerHTML)
```
seems, `editor` might be null here... in a very rare case

TomasSlama replied:
we have condition on line 14

dmaklygin1 commented on packages/picasso-lab/src/TextEditor/hooks/useHandleChangeFromController.tsx:
```diff
@@ -0,0 +1,34 @@
+  { value: controllerValue }: { value: Props['value'] }
+) => {
+  useEffect(() => {
+    if (editorRef?.current && typeof controllerValue === 'string') {
```
```
typeof controllerValue === 'string'
```
just thinking, how it's possible to pass different from `string` type here (`value` type is strictly defined as a string)

TomasSlama replied:
you are right, removed unnecessary condition.

dmaklygin1 commented on packages/picasso-lab/src/TextEditor/hooks/useHandleChangeFromController.tsx:
```diff
@@ -0,0 +1,34 @@
+        // @types/quill has incorrect convert typing
+        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+        // @ts-ignore
+        const delta = clipboard.convert(controllerValue)
```
What if controllerValue is longer than `maxlength` prop? 
Should we cut an extra text? 
I found a solution: https://github.com/quilljs/quill/issues/3009#issuecomment-618271044

TomasSlama replied:
yeah, we should, `maxlength` isn't implemented yet, I wasn't sure if we will need it and left it open for discussion, I think I will create another task for it.

dmaklygin1 commented on packages/picasso-lab/src/TextEditor/TextEditor.tsx:
```diff
@@ -0,0 +1,95 @@
+  /**
+   * Callback on text change
+   */
+  onChange: (value: HTMLString) => void
```
I wanted to talk about the definition of `onChange`: I would move out the definition of this callback, so the end developer may reuse it outside by effortlessly defining a callback, for ex:
```
export type TextEditorChangeHandler = (value: HTMLString) => void
```
and the implementation
```
const handleChange: TextEditorChangeHandler = () => ...
```


wdyt?

TomasSlama replied:
love it, thanks for hint ❤️

dmaklygin1 commented on packages/picasso-lab/src/TextEditor/story/Default.example.tsx:
```diff
@@ -0,0 +1,35 @@
+const Example = () => {
+  const [value, setValue] = useState<string | undefined>()
+
+  const handleChange = (newValue?: string) => setValue(newValue)
```
according to `onChange` type definition, newValue may not be `undefined`.
To avoid inconsistency, I would create a separate type for the onChange (like I highlighted above: TextEditorChangeHandler) and reuse it here.

TomasSlama replied:
✅

dmaklygin1 commented on packages/picasso-lab/src/TextEditor/story/Limit.example.tsx:
```diff
@@ -0,0 +1,36 @@
+    <>
+      <Container style={{ height: '300px' }}>
+        <TextEditor
+          id='limit'
```
what if we have two Components with the same ID. Whether they will work in integration?

TomasSlama replied:
The `id` must be unique, otherwise it would just reinit the previous one.

dmaklygin1 reviewed (commented):
Good work! Left some comments

OndrejTuma commented on packages/picasso-lab/src/TextEditor/hooks/useHandleChangeFromController.tsx:
```diff
@@ -0,0 +1,34 @@
+  }: { value: Props['value']; oldValue?: string }
+) => {
+  useEffect(() => {
+    if (editorRef?.current) {
```
Like @dmaklygin1 suggested, editorRef is mandatory here too, so no need for optional chaining here

OndrejTuma replied:
nit: here also may be more readable if you first deal with undefined quill like so:
`if (!editorRef.current) { return }`

TomasSlama replied:
✅

TomasSlama commented on packages/picasso-lab/src/TextEditor/hooks/useHandleAutofocus.tsx:
```diff
@@ -0,0 +1,22 @@
+  const firstRender = useRef(true)
+
+  useEffect(() => {
+    const editorApi = editorRef?.current
```
you are right, removed.

TomasSlama commented on packages/picasso-lab/src/TextEditor/hooks/useHandleChangeEvent.tsx:
```diff
@@ -0,0 +1,37 @@
+  useEffect(() => {
+    const { current: editor } = editorRef
+
+    if (editor) {
```
I don't know, I try to avoid negative conditions

OndrejTuma replied:
It's a nit, but I was promoting early return pattern which I like :-) feel free to resolve it

TomasSlama commented on packages/picasso-lab/src/TextEditor/hooks/useHandleChangeFromController.tsx:
```diff
@@ -0,0 +1,34 @@
+
+      // is value set from outside
+      if (editor && oldValue !== controllerValue) {
+        const delta = (editor.clipboard.convert as (html: string) => Delta)(
```
no need, because we have in top-level condition check if we have `controllerValue`

teimurjan commented on package.json:
```diff
@@ -98,6 +98,7 @@
     "happo-plugin-storybook": "^2.7.0",
     "happo.io": "^6.8.0",
     "hygen": "^6.0.4",
+    "identity-obj-proxy": "^3.0.0",
```
Why did you add it?

TomasSlama replied:
good point, it is leftover from some test. 

✅ removed.

teimurjan reviewed (commented):
[The problem](https://github.com/toptal/picasso/pull/2312/files#r765614496) @dmaklygin1 raised is not solved yet.

If we pass `editorRef.current` to all the hooks, we'll

1. Remove duplicated lines like `const editorApi = editorRef.current`
2. Solve the issue when the ref is changed but `useEffect` didn't run.

dmaklygin1 commented on packages/picasso-lab/src/TextEditor/hooks/useHandleChangeFromController.tsx:
```diff
@@ -0,0 +1,46 @@
+      const editor = editorRef.current
+      const currentValue = editor.root.innerHTML
+
+      const isValueChanged = currentValue !== controllerValue
```
I guess, this is an odd empty line :)

separatio reviewed (approved):
Manually tested. Looks good to go. :+1: 

Nit: http://localhost:9001/?path=/story/picasso-forms-form--form
The editor from forms does not have a placeholder text. I wonder if the validation works correctly while having a placeholder text as well.

Otherwise, as I said, :+1:

augustobmoura commented on packages/picasso/src/RichTextEditor/RichTextEditor.tsx:
```diff
@@ -0,0 +1,237 @@
+      dispatch
+    })
+
+    const defaultValueInHtml = useMemo(
+      // this effects needs to happen only once on first render
+      () =>
+        defaultValue
+          ? hastUtilToHtml(hastSanitize(defaultValue))
+          : defaultValue,
+      /* eslint-disable-next-line */
+      []
+    )
```
Instead of using eslint ignore comments we could refactor this to use an `useState` hook, it accepts a function as first argument that is only run when the hook is created. We can ignore the setter

```suggestion
    const [defaultValueInHtml] = useState(
      // this effects needs to happen only once on first render
      () =>
        defaultValue
          ? hastUtilToHtml(hastSanitize(defaultValue))
          : defaultValue
    )
```

OndrejTuma replied:
Agree, only input value should not be function in this case 👍

augustobmoura commented on packages/picasso/src/RichTextEditor/RichTextEditor.tsx:
```diff
@@ -0,0 +1,237 @@
+          } else if (ref != null) {
+            ref.current = node
+          }
+          // @ts-ignore
+          wrapperRef.current = node
```
We shouldn't use `@ts-ignore` **ever**, there are always better ways to do it. If the problem is that `wrapperRef.current` is readonly you could initialize `wrapperRef` as `useRef<HTMLDivElement | null>()` instead

TomasSlama replied:
✅

augustobmoura commented on packages/picasso/src/RichTextEditor/RichTextEditor.tsx:
```diff
@@ -0,0 +1,237 @@
+        />
+        <QuillEditor
+          ref={editorRef}
+          disabled={disabled!} // eslint-disable-line @typescript-eslint/no-non-null-assertion
```
We could just cast this variable to boolean and remove the non-null-assertion

```suggestion
          disabled={!!disabled}
```

TomasSlama replied:
✅

augustobmoura commented on packages/picasso/src/RichTextEditor/utils/convertFormat.ts:
```diff
@@ -0,0 +1,38 @@
+} from '../../QuillEditor'
+import { FormatType as ToolbarFormatType, HeaderValue } from '../store/toolbar'
+
+export const convertBoldFromEditorValue = (bold: EditorBoldValue) =>
+  bold || false
+export const convertItalicFromEditorValue = (italic: EditorItalicValue) =>
+  italic || false
+export const convertListFromEditorValue = (list: EditorListValue) =>
+  list || false
```
What's the point of these functions?

TomasSlama replied:
@augustobmoura Quill internally shows only active format, so when we get format from Quill.getFormat() it is `{}` by default. When you have bold active `{ bold: true }`, when toggle, back to `{}`. So we transform here `undefined` to `false`

augustobmoura replied:
Yeah, but they seem kinda pointless, their whole implementation could be replaced by `!!`, doesn't look like something that needs to be a special function, you could just inline it where is being called. I think we can postpone their removal to future ticket, your call

augustobmoura commented on tsconfig.json:
```diff
@@ -34,6 +34,6 @@
     ]
   },
-  "include": ["packages/*/src"],
+  "include": ["packages/*/src", "packages/picasso/RichTextEditor"],
```
I don't think this line is needed anymore

TomasSlama replied:
✅

ozgurkececioglu commented on cypress/component/RichTextEditor.spec.tsx:
```diff
@@ -0,0 +1,365 @@
+
+  it('handles keybindings correctly', () => {
+    // render the editor
+    mount(renderEditor(defaultProps))
+    setAliases()
```
Couldn't we use beforeEach for these initialization parts?

ozgurkececioglu commented on cypress/component/RichTextEditor.spec.tsx:
```diff
@@ -0,0 +1,365 @@
+      cy.get('@olButton').should('not.have.attr', 'disabled')
+    })
+
+    it.only('does not open selectbox on click when toolbar is disabled', () => {
```
Is `only` needed still?

TomasSlama replied:
✅

ozgurkececioglu commented on packages/picasso-forms/src/Form/story/index.jsx:
```diff
@@ -218,6 +218,10 @@ however, you may need custom validators for more complex types of fields.
   }) // picasso-skip-visuals
+  .addExample('Form/story/RichTextEditor.example.tsx', {
+    title: 'Rich text editor',
+    description: 'Showcase how to display field requirements.'
```
Copy/paste error?

TomasSlama replied:
✅

ozgurkececioglu commented on packages/picasso/src/QuillEditor/hooks/useQuillInstance/test.tsx:
```diff
@@ -0,0 +1,47 @@
+    const id = 'editor'
+
+    // quill needs to have container for both toolbar and editor
+    // to successfuly initit
```
```suggestion
    // to successfuly init
```

TomasSlama replied:
✅

ozgurkececioglu commented:
> When I visit texteditor component in picasso it jumps to the second example, is this intended?
> 
> ![simplescreenrecorder-2021-12-09_13 44 17](https://user-images.githubusercontent.com/72510037/145372972-b5182c3a-1dde-49da-983d-0bbe2348a9da.gif)

As Lasha pointed out before `autoFocus` is active for both RichtextEditor pages and Form stories. These pages are directly jumping to RichtextEditor which is not a good idea for storybook.

TomasSlama commented on __tests__/styleMock.js:
```diff
@@ -0,0 +1 @@
+module.exports = {}
```
we use it in `jest.config`. It was necessary, we were importing CSS file directly. We don't anymore, so we don't need it anymore, but it might still be useful in future 🤔, when someone will need to use CSS file. 🤷

augustobmoura replied:
Ok, I think we can leave it for now

TomasSlama commented on packages/picasso/src/RichTextEditor/hooks/useCounter/test.ts:
```diff
@@ -0,0 +1,97 @@
+    const hookOptions = {
+      maxLength: 8,
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
+      maxLengthMessage: (_maxLength: number, _currLength: number) =>
```
✅

OndrejTuma commented on .changeset/forty-paws-deliver.md:
```diff
@@ -0,0 +1,10 @@
+---
+'@toptal/picasso': minor
+'@toptal/picasso-forms': minor
```
Only minor? 🙂

TomasSlama commented:
@LashaJini @OndrejTuma @ozgurkececioglu I have fixed the unexpected autoFocus of the second example. The issue was in `useDefaultValue` triggering `quill.setContents` with `api` event type. It seems that `api` event automatically focuses on the Editor. So I changed it to `user` and everything seems to work as expected. Thanks for pointing this issue out! 🙇

OndrejTuma commented:
> @LashaJini @OndrejTuma @ozgurkececioglu I have fixed the unexpected autoFocus of the second example. The issue was in `useDefaultValue` triggering `quill.setContents` with `api` event type. It seems that `api` event automatically focuses on the Editor. So I changed it to `user` and everything seems to work as expected. Thanks for pointing this issue out! 🙇

Its still happening to me 🙁 on latest [https://picasso.toptal.net/FX-2293-text-editor/](https://picasso.toptal.net/FX-2293-text-editor/?a14c771552e45f385643bbc1f63d24c99236c5f6)

ozgurkececioglu commented:
> > @LashaJini @OndrejTuma @ozgurkececioglu I have fixed the unexpected autoFocus of the second example. The issue was in `useDefaultValue` triggering `quill.setContents` with `api` event type. It seems that `api` event automatically focuses on the Editor. So I changed it to `user` and everything seems to work as expected. Thanks for pointing this issue out! 🙇
> 
> Its still happening to me 🙁 on latest [https://picasso.toptal.net/FX-2293-text-editor/](https://picasso.toptal.net/FX-2293-text-editor/?a14c771552e45f385643bbc1f63d24c99236c5f6)

It's fixed for Form section but it is still happening for me too for this page: https://picasso.toptal.net/FX-2293-text-editor/?path=/story/components-richtexteditor--richtexteditor

augustobmoura commented on packages/picasso/src/RichTextEditor/RichTextEditor.tsx:
```diff
@@ -0,0 +1,230 @@
+    })
+
+    const [defaultValueInHtml] = useState(
+      defaultValue ? hastUtilToHtml(hastSanitize(defaultValue)) : defaultValue
```
Because we are not wrapping the default in a function `hastUtilToHtml` is being run every re-render. React already supports the first argument as a function and only run it on creation
```suggestion
      () => defaultValue ? hastUtilToHtml(hastSanitize(defaultValue)) : defaultValue
```

TomasSlama replied:
nice, didn't know it.

augustobmoura commented:
LGMT! :rocket:

OndrejTuma reviewed (approved):
Everything works 🎉
